File: tetrattack_generator/t_generator.hs

module TetrattackGenerator (
) where
import TetrattackBoard
import System.Random
import Data.Maybe

type Line = [Pos]

-- creates every possible combination of 2 lists
permute_zip :: [a] -> [a] -> [(a, a)]
permute_zip list1 list2 = concat (map (\element1 -> map (\element2 -> (element1, element2)) list2) list1)

-- my random until i can figure out proper syntax for this thing
type Randoms = ([Bool], [Int])

random_bool :: Randoms -> (Randoms, Bool)
random_bool (bool:bools, ints) = ((bools, ints), bool)

random_int :: Randoms -> (Randoms, Int)
random_int (bools, int:ints) = ((bools, ints), int)

random_int_range :: Randoms -> Int -> Int -> (Randoms, Int)
random_int_range randoms min max = (new_randoms, (int `mod` (max - min + 1)) + min)
    where (new_randoms, int) = random_int randoms

random_element :: Randoms -> [a] -> (Randoms, a)
random_element randoms elements = (new_randoms, elements !! index)
    where (new_randoms, index) = random_int_range randoms 0 ((length elements) - 1)
    
generate_next_thing :: Randoms -> Board -> (Randoms, Action)
generate_next_thing randoms board
  | action_type == 0 = generate_next_create new_randoms board
  | otherwise = generate_next_move new_randoms board
  where (new_randoms, action_type) = random_bool randoms

line_is_supported :: Board -> Line -> Bool
line_is_supported board [] = True
line_is_supported board ((x,y):line) = pos_is_supported && line_is_supported (board_insert (Just Red) (x,y) board) line
    where pos_is_supported = y == 0 || isJust (board_cell (x,y-1) board)

-- generate a new line
--  horz or vert
--  position
--  color
--  length (todo: combos?)
-- make sure it's valid -- don't create a horz floating in space
-- generate a whole ton of them and pick the first good one
generate_next_create :: Randoms -> Board -> (Randoms, Action)
generate_next_create randoms board = valid_candidate randoms
    where (randoms2, color) = random_element randoms [Red, Yellow, Blue, Green, Pink, Black, Orange]
          valid_candidate randoms = if (line_is_supported board line)
                                        then (randoms2, create_cells color line)
                                        else valid_candidate randoms2
            where (randoms2, line) = candidate randoms
          candidate randoms = (randoms5, line)
            where (randoms2, length) = random_int_range randoms 3 5
                  (randoms3, x) = random_int_range randoms2 0 ((board_num_columns board) - length)
                  (randoms4, y) = random_int_range randoms3 0 ((board_num_rows board) - length)
                  (randoms5, is_vert) = random_bool randoms4
                  line = if is_vert
                            then generate_line (\(x, y) -> (x, y + 1)) color length (x,y)
                            else generate_line (\(x, y) -> (x + 1, y)) color length (x,y)

generate_line :: (Pos -> Pos) -> Color -> Int -> Pos -> Line
generate_line_vert _ _ 0 _ = []
generate_line next_func color length pos = pos : generate_line next_func color (length - 1) (next_func pos)

-- generate every possible move
-- then, pick one
generate_next_move :: Randoms -> Board -> (Randoms, Action)
generate_next_move random board = random_element new_random all_moves
  where (new_random, all_moves) = generate_all_moves board

generate_all_moves :: Board -> [Action]
generate_all_moves board = concat (map (generate_all_cell_moves board) poses)
  where poses = permute_zip [0..x_max-1] [0..y_max-1]
        x_max = board_num_rows board
        y_max = board_num_columns board

generate_all_cell_moves :: Board -> Pos -> [Action]
generate_all_cell_moves board (x, y)
  | Nothing == board_cell (x, y) board = []
  | otherwise = concat (column_actions (x-1, y) ++ column_actions (x+1, y))
    where column_actions column =
            if column > 0 && column < x_max
                then board_swap ((x, y), (column, y)) : ((\board -> climb_for_inserts board (x, y) (column, y)) board)
                else []
          x_max = board_num_rows board
          y_max = board_num_columns board

climb_for_inserts :: Board -> Pos -> Pos -> [Action]
climb_for_inserts board source (dest_x, dest_y)
  | Nothing == board_cell (dest_x, dest_y) board = []
  | otherwise = board_move_and_insert source (dest_x, dest_y) : climb_for_inserts board source (dest_x, dest_y + 1)

create_cells :: Color -> Line -> Action
create_cells color [] = no_action
create_cells color (pos:poses) = (\board -> create_cells color poses (board_after_inserting_one board))
    where board_after_inserting_one board = board_insert (Just color) pos board

File: tetrattack_generator/js/t_generator.js


// return an action that could be applied to this board to generate stuff
// Chooses randomly!
function generate_next_thing(board) {
    // keep generating until we make one that doesn't suck
    var action = null;
    while (action === null) {
        const matches = board_get_matches(board);
        if (matches.length) {
            action = generate_next_move(matches);
        } else {
            if (random_bool()) {
                action = generate_next_move();
            } else {
                action = generate_next_create(board);
            }
        }
    }
    return action;
}

function board_cell_matches(board, pos, iterate_function) {
    const color = board.cells.get(pos);
    const match_length = 3;
    let iterator = pos;
    
    for (let i = 0; i < match_length; ++i) {
        iterator = iterate_function(iterator);
        if (board.cells.get(iterator) !== color) return false;
    }
    
    return true;
}

function board_get_matches(board) {
    const matches = [];
    const match_length = 3;
    
    for (let x = 0; x < board.num_columns; ++x) {
        for (let y = 0; y < board.num_rows; ++y) {
            const pos = {x, y};
            if (board_cell_matches(board, pos, pos_above) || board_cell_matches(board, pos, pos_right)) {
                matches.push(pos);
            }
        }
    }

    return matches;
}

function line_is_supported(board, line) {
    switch (line.isVert) {
        case true: return cell_is_supported(board, line.pos);
        case false: return line_cells(line).all(function (pos) { return cell_is_supported(board, pos); });
    }
}

function cell_is_supported(board, pos) {
    return pos.y === 0 || board.cells.get(pos_below(pos));
}

/*
-- generate a new line
--  horz or vert
--  position
--  color
--  length (todo: combos?)
-- make sure it's valid -- don't create a horz floating in space
-- generate a whole ton of them and pick the first good one
*/

/*
 * line:
 * {
 *     pos: pos,
 *     color: color,
 *     isVert: bool,
 *     length: int,
 *     cells(): [pos]
 * }
 */

function line_cells(line) {
    const result = [];
    let pos = line.pos;
    for (let i = 0; i < line.length; ++i) {
        result.push(pos);
        if (isVert) {
            pos = pos_above(pos);
        } else {
            pos = pos_right(pos);
        }
    }
    return result;
}

function generate_next_create(board) {
    const color = random_element(colors);
    const length = random_int_inclusive(3, 5);
    const x = random_int_exclusive(0, board.num_columns);
    const y = random_int_exclusive(0, board.num_rows);
    const isVert = random_bool();
    const line = {
        pos: {x, y},
        color,
        isVert,
        length
    };
    if (line_is_supported(board, line)) {
        return line;
    } else {
        return null;
    }
}

function generate_next_move(board) {
    const pos = random_element(board.cells).keys;

    if (!board.cells.get(pos)) {
        return null;
    }

    const candidates = generate_next_move_candidates(board, pos, pos);
    if (candidates && candidates.length) {
        const action = random_element(candidates);
        if (!board_get_matches(apply_action(board, action)).length) {
            return null;
        } else {
            return action;
        }
    } else {
        return null;
    }
}

function generate_next_move_candidates(board, pos, orig_pos) {
    // 2 kinds of move:
    const color = board.cells.get(pos);
    const results = [];
    
    // 1) swap with a full peer (swap)
    // this can't happen when we're above the pos (color == null protects us from this)
    
    if (color) {
        if (pos.x > 0 && board.cells.get(pos_left(pos)) !== color) {
            results.push(prepare_swap(pos_left(pos), pos));
        }

        if (pos_right(pos.x) < board.num_columns && board.cells.get(pos_right(pos)) !== color) {
            results.push(prepare_swap(pos, pos_right(pos)));
        }
    }
    
    // 2) move stuff from a column beside
    // in reverse this looks like taking the orig cell, maybe moving it up, then pushing it left or right (and shuffling other stuff up)
    if (pos.x > 0 && cell_is_supported(board, pos_left(pos))) {
        results.push(prepare_move(orig_pos, pos_left(pos)));
    }

    if (pos_right(pos.x) < board.num_columns && cell_is_supported(board, pos_right(pos))) {
        results.push(prepare_move(orig_pos, pos_right(pos)));
    }
    
    if (results.length && !board.cells.get(pos_above(pos))) {
        return results ++ generate_next_move_candidates(board, pos_above(pos), orig_pos);
    } else {
        return results;
    }
}
